### PR TITLE
Extend the harness for priority queue s_swap with fc assertions

### PR DESCRIPTION
**Pull request by angelhof**
Extend the harness for priority queue s_swap with functional correctness assertions.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: source/priority_queue.c [hunk 1]
```diff
@@ -206,6 +210,14 @@ bool aws_priority_queue_is_valid(const struct aws_priority_queue *const queue) {
     return pred_is_valid && container_is_valid && backpointer_list_is_valid && backpointer_struct_is_valid;
 }
 
+bool aws_backpointer_index_valid(const struct aws_priority_queue *const queue, size_t index) {
+    if (queue->backpointers.data && index < queue->backpointers.length) {
+        struct aws_priority_queue_node *node = ((struct aws_priority_queue_node **)queue->backpointers.data)[index];
```
Comment (karkhaz) [inline source/priority_queue.c]: Is the cast to `aws_priority_queue_node **` (rather than just single `*`) correct?

Comment (angelhof) [reply source/priority_queue.c]: Thanks for the comment. From my understanding the backpointers array list holds pointers to `struct aws_priority_queue_node`. So the backpointers.data should be a double pointer. Does this sound reasonable?

**Comment by karkhaz** [comment]
@angelhof please rebase so that this is mergeable. I've assigned c-common people for review.

**Comment by angelhof** [comment]
There seems to be a problem with the s_sift_either proof (even though I didn't directly change it). It seems to be stuck on CBMC batch. I have to debug before this is merged.

**Comment by angelhof** [comment]
The downstream build error seems to be unrelated to these changes as it happens in #421 too. I also tried to run the downstream build on master locally and it fails at the same place.

File: source/priority_queue.c [hunk 2]
```diff
@@ -198,12 +222,33 @@ bool aws_priority_queue_is_valid(const struct aws_priority_queue *const queue) {
     bool backpointer_list_item_size = queue->backpointers.item_size == sizeof(struct aws_priority_queue_node *);
     bool lists_equal_lengths = queue->backpointers.length == queue->container.length;
     bool backpointers_non_zero_current_size = queue->backpointers.current_size > 0;
+
+    /* This check must be guarded, as it is not efficient, neither
+     * when running tests nor CBMC */
+#ifdef PQ_BACKPOINTERS_CHECK
```
Comment (danielsn) [inline source/priority_queue.c]: Is this the best name here?

File: source/priority_queue.c [hunk 3]
```diff
@@ -180,15 +184,35 @@ void aws_priority_queue_init_static(
...
 
-bool aws_priority_queue_is_valid(const struct aws_priority_queue *const queue) {
-    /* Pointer validity checks */
+bool aws_priority_queue_backpointer_index_valid(const struct aws_priority_queue *const queue, size_t index) {
+    if (queue->backpointers.data == NULL) {
+        return true;
+    }
+    if (index < queue->backpointers.length) {
+        struct aws_priority_queue_node *node = ((struct aws_priority_queue_node **)queue->backpointers.data)[index];
+        return node == NULL || AWS_MEM_IS_WRITABLE(node, sizeof(struct aws_priority_queue_node));
```
Comment (danielsn) [inline source/priority_queue.c]: Use brackets

File: source/priority_queue.c [hunk 4]
```diff
@@ -198,12 +222,33 @@ bool aws_priority_queue_is_valid(const struct aws_priority_queue *const queue) {
     bool backpointer_list_item_size = queue->backpointers.item_size == sizeof(struct aws_priority_queue_node *);
     bool lists_equal_lengths = queue->backpointers.length == queue->container.length;
     bool backpointers_non_zero_current_size = queue->backpointers.current_size > 0;
+
+    /* This check must be guarded, as it is not efficient, neither
+     * when running tests nor CBMC */
+#ifdef PQ_BACKPOINTERS_CHECK
+    bool backpointers_allocated = aws_priority_queue_backpointers_valid_deep(queue);
```
Comment (danielsn) [inline source/priority_queue.c]: maybe call the function call _allocated? Or change the variable name.

File: source/priority_queue.c (hunk 2 above)
Comment (angelhof) [reply source/priority_queue.c]: I changed that name to `AWS_DEEP_CHECKS`, which should be used to guard any validity check that is not O(1).

**Pull request merged**